### PR TITLE
NAS-106293 / 12.1 / Add more stats to usage plugin

### PR DESCRIPTION
**Pull request by sonicaj**

### CODE REVIEW

**Comment by william-gr** [comment]
@sonicaj wrong base branch?

**Comment by sonicaj** [comment]
That's deliberate, as it uses the changes introduced in share/tasks PR. Once that's merged, I'll change to master :)

File: src/middlewared/middlewared/service.py [hunk 1]
```diff
@@ -624,6 +624,38 @@ async def generate_locked_alert(self, share_task_id):
             {**share_task, 'identifier': await self.human_identifier(share_task), 'type': self.share_task_type}
         )
 
+    @private
+    def datasets_in_use(self, share_task_id, datasets=None):
```
Comment (william-gr) [inline src/middlewared/middlewared/service.py:628]: It seems to be we are starting to overcomplicate this service base, just for some obscure use case?

Comment (william-gr) [reply src/middlewared/middlewared/service.py:628]: How about libzfs.get_dataset_by_path ?

File: src/middlewared/middlewared/plugins/zfs.py [hunk 2]
```diff
@@ -501,6 +501,28 @@ def common_encryption_checks(self, ds):
...
+            datasets = [
+                d['id'] for d in
+                self.middleware.call_sync(
+                    'zfs.dataset.query', [], {
+                        'select': ['id'], 'extra': {'properties': [], 'user_properties': False}
+                    }
+                )
+            ]
+        if not (path or '').startswith('/mnt'):
+            return
```
Comment (william-gr) [inline src/middlewared/middlewared/plugins/zfs.py:515]: this should come before datasets query

File: src/middlewared/middlewared/plugins/zfs.py [hunk 3]
```diff
@@ -501,6 +501,28 @@ def common_encryption_checks(self, ds):
         if not ds.encrypted:
             raise CallError(f'{id} is not encrypted')
 
+    def get_dataset_by_path(self, path, datasets=None):
```
Comment (william-gr) [inline src/middlewared/middlewared/plugins/zfs.py:504]: I dont understand why we cant use libzfs.get_dataset_by_path ?

File: src/middlewared/middlewared/plugins/usage.py [hunk 4]
```diff
@@ -46,34 +49,82 @@ async def start(self):
...
-                **jails,
-                **network,
-                **{k: v for l in system['gather_system'] for k, v in l.items()},
-                **plugins,
-                **pools,
-                **services,
-                **sharing,
-                **vms
+                k: v for f in dir(self) if f.startswith('gather_') and callable(getattr(self, f))
+                for k, v in (self.middleware.call_sync(f'usage.{f}', context)).items()
```
Comment (william-gr) [inline src/middlewared/middlewared/plugins/usage.py:64]: There is an extra parenthesis here we can drop?

File: src/middlewared/middlewared/plugins/usage.py [hunk 5]
```diff
@@ -46,34 +49,82 @@ async def start(self):
...
-        services = await self.gather_services()
-        sharing = await self.gather_sharing()
-        vms = await self.gather_vms()
+    def gather(self):
+        datasets = self.middleware.call_sync('zfs.dataset.query')
+        context = {
+            'network': self.middleware.call_sync('interfaces.query'),
+            'root_datasets': {d['id']: d for d in datasets if '/' not in d['id']},
+            'zvols': filter_list(datasets, [['type', '=', 'VOLUME']]),
+            'datasets': {d['id']: d for d in datasets},
```
Comment (william-gr) [inline src/middlewared/middlewared/plugins/usage.py:58]: We are iterating datasets two times, seems we could:
```
for d in self.middleware.call_sync('zfs.dataset.query'):
   context['datasets'][d['id']] = d
   if '/' not in d['id']:
       context['root_datasets'] = d
```
or something like that?

File: src/middlewared/middlewared/plugins/usage.py [hunk 6]
```diff
@@ -46,34 +49,82 @@ async def start(self):
...
+        datasets_data = context['datasets']
+        datasets = deepcopy(datasets_data)
+        for namespace in ('cloudsync', 'rsynctask'):
+            task_datasets = deepcopy(datasets_data)
+            for task in self.middleware.call_sync(
+                f'{namespace}.query', [['enabled', '=', True], ['direction', '=', 'PUSH'], ['locked', '=', False]]
+            ):
+                task_ds = self.middleware.call_sync('zfs.dataset.get_dataset_by_path', task['path'], task_datasets)
+                if task_ds:
+                    backed[namespace] += task_datasets.pop(task_ds)['properties']['used']['parsed']
```
Comment (william-gr) [inline src/middlewared/middlewared/plugins/usage.py:84]: How do you know task_datasets has not been popped before for task_ds?

File: src/middlewared/middlewared/plugins/zfs.py (hunk 3 above)
Comment (sonicaj) [reply src/middlewared/middlewared/plugins/zfs.py:504]: It wouldn't work with say path `/mnt/test/abc` if the only dataset we have is `test`

File: src/middlewared/middlewared/plugins/usage.py (hunk 6 above)
Comment (sonicaj) [reply src/middlewared/middlewared/plugins/usage.py:84]: `get_dataset_by_path` ensures we found the dataset in `task_datasets`

File: src/middlewared/middlewared/plugins/zfs.py (hunk 3 above)
Comment (william-gr) [reply src/middlewared/middlewared/plugins/zfs.py:504]: Are you sure?

truenas# mkdir /mnt/tank/foo/bar
truenas# python3
Python 3.8.3rc1 (default, Apr 30 2020, 07:33:30) 
[GCC 9.3.0] on linux
Type "help", "copyright", "credits" or "license" for more information.
>>> import libzfs
>>> libzfs.ZFS().get_dataset_by_path
<built-in method get_dataset_by_path of libzfs.ZFS object at 0x7fc8f1da3bd0>
>>> libzfs.ZFS().get_dataset_by_path('/mnt')
<libzfs.ZFSDataset name 'boot-pool/ROOT/13.0-MASTER-20200625-133242' type 'FILESYSTEM'>
>>> libzfs.ZFS().get_dataset_by_path('/mnt/tank')
<libzfs.ZFSDataset name 'tank' type 'FILESYSTEM'>
>>> libzfs.ZFS().get_dataset_by_path('/mnt/tank/foo')
<libzfs.ZFSDataset name 'tank/foo' type 'FILESYSTEM'>
>>> libzfs.ZFS().get_dataset_by_path('/mnt/tank/foo/bar')
<libzfs.ZFSDataset name 'tank/foo' type 'FILESYSTEM'>

Comment (sonicaj) [reply src/middlewared/middlewared/plugins/zfs.py:504]: Interesting
```
>>> import libzfs;z=libzfs.ZFS()
>>> z.get_dataset_by_path('/mnt/test')
<libzfs.ZFSDataset name 'test' type 'FILESYSTEM'>
>>> z.get_dataset_by_path('/mnt/test/abc')
cannot open '/mnt/test/abc': No such file or directory
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "libzfs.pyx", line 1199, in libzfs.ZFS.get_dataset_by_path
libzfs.ZFSException: Dataset with path /mnt/test/abc not found
>>>
```
I'll look closely :)

Comment (sonicaj) [reply src/middlewared/middlewared/plugins/zfs.py:504]: Oh, I think I found the issue :) My bad, thank you for pointing it out

**Pull request merged**